### DISCUSSION TITLE
Has anyone done an index and concordance yet?

### DISCUSSION THREAD

**Discussion by anythingwithawire** [markdoc/markdoc#96] [Ideas]
I would really like to be able to generate a hyperlinked index and a concordance (like an index but shows a certain number of words either side of the index term to show context, as well as possible the nested heading structure context).

The words or phrases to index and concord could be from a seperate file input, be based on a glossary embedded in the doc (this could be super useful) or maybe extracted based on a regex string.

While wishing, the ability to span single or multiple docs would be super handy.

Has any one already done anything like this and can give examples?

**Comment by nkohari-stripe** [comment]
Love this idea!

We haven't done anything quite like this at Stripe, yet! The closest we've implemented is a glossary, which uses a tag like:

```
{% glossary term="PCI Compliant" /%}
```

This tag is rendered into an interactive link which, when clicked, shows a popover with the definition of the term. The definitions are stored in a separate YAML file. In practice, it looks like this:

![image](https://user-images.githubusercontent.com/53919368/174866650-8cd001b2-713d-43a8-84d3-9f2346915e44.png)

Anyhow, building an index should definitely be feasible. Given a list of words that you'd want to index, you could walk the Markdoc AST during your build for each document in your corpus looking for `text` nodes which contain the word. Then, you could keep a list of documents which contain the word.

However... linking from the index page directly to the location _within_ a document where the word occurs might be a little more challenging, since you'd really need an `<a>` tag for the browser to target. To solve this, when you find a node which has a reference to an indexed word, you could transform the AST to replace the word with a specific Markdoc tag. Then, that tag could be the target for the link in the index.

You've piqued my curiosity, and I may try to implement a simple version of this. I'll report back if I'm able to do so!

**Comment by bwklein** [comment]
Indexes are often manually built by tagging specific locations in a document for a term in the index to be anchored, to where not every use of the word anywhere in the document should be linked to in the Index.

@nkohari-stripe I think that your Glossary example is actually closer to something I would want to do.

`{% index words="PCI Compliant" parent="Security" /%}`

In that location within the document, there would be the 'words' as text, with an anchor next to it for where the links in the Index should href.

Then in the 'Index' section of the document, there would be an alphanumeric list of all 'parents' with nested 'words' and/or 'words' at the top level sort, that have been tagged in the documents.

**Reply by mfix-stripe** [reply]
Thanks @bwklein! (Cc @anythingwithawire for visibility)
